spdx-utils/src/main/kotlin/SpdxExpression.kt

```diff
@@ -243,6 +243,17 @@ data class SpdxCompoundExpression(
             SpdxOperator.AND -> left.offersChoice() || right.offersChoice()
         }
 
+    override fun equals(other: Any?): Boolean {
+        if (other !is SpdxExpression) return false
+
+        val validChoices = validChoices().map { it.decompose() }
+        val otherValidChoices = other.validChoices().map { it.decompose() }
+
+        return validChoices.size == otherValidChoices.size && validChoices.all { otherValidChoices.contains(it) }
+    }
```

Comment:
I have difficulties understanding why `equals()` is overridden multiple times, in different subclasses of SPDX expression. Would it work out to place this equals implementation in the `SpdxExpression` class so that all other `equals()` overrides would become unnecessary?

Reply:
If `equals()` would be implemented in `SpdxExpression` it would require checking properties of it's subclasses, e.g. to check if the `id` property of two `SpdxLicenseIdExpression`s is equal. It would also become a lot more nested, as the different handling for each subclass would have to be done in a single method.

It's like with the classic inheritance example with different shapes inheriting from a `Shape` class, would you try to do all comparisons in `Shape.equals()` or make separate functions like `Circle.equals()` and `Rectangle.equals()`?

Reply:
> If equals() would be implemented in SpdxExpression it would require checking properties of it's subclasses, e.g. to check if the id property of two SpdxLicenseIdExpressions is equal. 

Why would that be necessary? The implementation here compares only the return value from `validChoices()`, and as `validChoices()` is implemented / inherited by all subtypes there is no casting involved. I guess  I'm missing something again?

Reply:
Right, thanks!